utils/common.sh

```diff
@@ -193,6 +193,58 @@ function get_os_information() {
     echo -e "[$done_format]"
 }
 
+
+apt_ensure(){
+    __doc__="
+    Checks to see if the packages are installed and installs them if needed.
+
+    The main reason to use this over normal apt install is that it avoids sudo
+    if we already have all requested packages.
+
+    Args:
+        *ARGS : one or more requested packages
+
+    Environment:
+        UPDATE : if this is populated also runs and apt update
+
+    Example:
+        apt_ensure git curl htop
+    "
+    # Note the $@ is not actually an array, but we can convert it to one
+    # https://linuxize.com/post/bash-functions/#passing-arguments-to-bash-functions
+    ARGS=("$@")
+    MISS_PKGS=()
+    HIT_PKGS=()
+    _SUDO=""
+    if [ "$(whoami)" != "root" ]; then
+        # Only use the sudo command if we need it (i.e. we are not root)
+        _SUDO="sudo "
+    fi
+    # shellcheck disable=SC2068
+    for PKG_NAME in ${ARGS[@]}
+    do
+        # Check if the package is already installed or not
+        if dpkg -l "$PKG_NAME" | grep "^ii *$PKG_NAME" > /dev/null; then
+            echo "Already have PKG_NAME='$PKG_NAME'"
+            # shellcheck disable=SC2268,SC2206
+            HIT_PKGS=(${HIT_PKGS[@]} "$PKG_NAME")
+        else
+            echo "Do not have PKG_NAME='$PKG_NAME'"
+            # shellcheck disable=SC2268,SC2206
+            MISS_PKGS=(${MISS_PKGS[@]} "$PKG_NAME")
+        fi
+    done
+    # Install the packages if any are missing
+    if [ "${#MISS_PKGS}" -gt 0 ]; then
+        if [ "${UPDATE}" != "" ]; then
+            $_SUDO apt update -y
+        fi
+        DEBIAN_FRONTEND=noninteractive $_SUDO apt install --no-install-recommends -y "${MISS_PKGS[@]}"
+    else
+        echo "No missing packages"
+    fi
+}
```

Comment:
I think the original is ok. What is the problem with it? Can you provide an example where it fails?

Reply:
Upon further consideration, I like the AI version better. It makes sense to use the more robust version of array length that can handle arguments with spaces in case someone uses this code as an example for something where spaces can exist. Can you redo the suggestion so I can apply it?